shapeshift/web
# feat: manual receive address

## Description

- Adds the ability to manually enter a receive address when trading into an asset that is not supported by the connected wallet (e.g. ETH into DOGE on MetaMask)
- Validates that the receive asset is valid on the receiving chain
- Supports yatt addresses when the receiving chain is ETH (not a case we currently require, but it's there nonetheless)
- Supports ENS/unstoppable domain resolution

First successful TX with manual receive address: https://viewblock.io/thorchain/tx/1C9A199669FCD8D21F2EE20EFA91A6A20D5183A84749A96795E8F54BCC2A6C57

## Pull Request Type


## Issue (if applicable)

Closes https://github.com/shapeshift/web/issues/2168

## Risk

🌶️ Very high. 

It's a simple PR in many ways, but getting this wrong could mean loss of funds for the user if the funds are send to the wrong/invalid address.

## Testing

- Test sending into assets not supported by the connected wallet and ensure that the funds are received into the specified address.
- Ensure invalid addresses cannot be entered
- Ensure both ENS and unstoppable domain addresses work as expected

Try break it. Switch between supported and non-supported receive assets. Enter valid and invalid addresses.

Also sanity check regular asset sends still work, as I modified the send input component to make it more generic.

### Engineering

☝️

- Check that the `receiveAddress` in the Swapper Redux store is _always_ exactly what we expect.
- Check that the quote and trade requests are using the correct address in the "destination" address.

This is not a feature we can afford to get wrong - scrutinise the code accordingly.

### Operations

☝️

## Screenshots (if applicable)

<img width="473" alt="Screenshot 2023-05-11 at 3 17 33 pm" src="https://github.com/shapeshift/web/assets/97164662/a8360097-1616-45ba-8e0f-0ee5b6dee7c7">

<img width="480" alt="Screenshot 2023-05-11 at 3 19 41 pm" src="https://github.com/shapeshift/web/assets/97164662/ceeab702-cf5c-42da-9e98-b58288b53c71">

<img width="480" alt="Screenshot 2023-05-11 at 3 20 14 pm" src="https://github.com/shapeshift/web/assets/97164662/af24182f-7bf1-4b4d-b71f-8f2252bfde33">
